app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.ts

```diff
@@ -0,0 +1,265 @@
+  > = pick(requestedPermissions, [
+    RestrictedMethods.eth_accounts,
+    PermissionNames.permittedChains,
+  ]);
+  delete requestedPermissions[RestrictedMethods.eth_accounts];
+  delete requestedPermissions[PermissionNames.permittedChains];
+
+  // We manually handle eth_accounts and permittedChains permissions
+  // by calling the ApprovalController rather than the PermissionController
+  // because these two permissions do not actually exist in the Permssion
+  // Specifications. Calling the PermissionController with them will
+  // cause an error to be thrown. Instead, we will use the approval result
+  // from the ApprovalController to form a CAIP-25 permission later.
+  let legacyApproval;
+  const haveLegacyPermissions =
+    Object.keys(legacyRequestedPermissions).length > 0;
+  if (haveLegacyPermissions) {
+    if (!legacyRequestedPermissions[RestrictedMethods.eth_accounts]) {
+      legacyRequestedPermissions[RestrictedMethods.eth_accounts] = {};
+    }
+
+    if (!legacyRequestedPermissions[PermissionNames.permittedChains]) {
+      legacyRequestedPermissions[PermissionNames.permittedChains] = {};
+    }
+
+    if (isSnapId(origin)) {
+      delete legacyRequestedPermissions[PermissionNames.permittedChains];
+    }
+
+    legacyApproval = await requestPermissionApprovalForOrigin(
+      legacyRequestedPermissions,
+    );
+  }
+
+  let grantedPermissions: GrantedPermissions = {};
+  // Request permissions from the PermissionController for any permissions other
+  // than eth_accounts and permittedChains in the params. If no permissions
+  // are in the params, then request empty permissions from the PermissionController
+  // to get an appropriate error to be returned to the dapp.
+  if (
+    (Object.keys(requestedPermissions).length === 0 &&
+      !haveLegacyPermissions) ||
+    Object.keys(requestedPermissions).length > 0
+  ) {
+    const [_grantedPermissions] = await requestPermissionsForOrigin(
+      requestedPermissions,
+    );
+    // permissions are frozen and must be cloned before modified
+    grantedPermissions = { ..._grantedPermissions };
+  }
+
+  if (legacyApproval) {
+    let newCaveatValue = {
+      requiredScopes: {},
+      optionalScopes: {},
+      isMultichainOrigin: false,
+    };
+
+    if (isSnapId(origin)) {
+      newCaveatValue.optionalScopes = {
+        'wallet:eip155': {
+          accounts: [],
+        },
+      };
+    } else {
+      newCaveatValue = setPermittedEthChainIds(
+        newCaveatValue,
+        legacyApproval.approvedChainIds,
+      );
+    }
+
+    newCaveatValue = setEthAccounts(
+      newCaveatValue,
+      legacyApproval.approvedAccounts,
+    );
+
+    const permissions = getPermissionsForOrigin() || {};
+    let caip25Endowment = permissions[Caip25EndowmentPermissionName];
+    const existingCaveatValue = caip25Endowment?.caveats?.find(
+      ({ type }) => type === Caip25CaveatType,
+    )?.value as Caip25CaveatValue;
```

Comment:
For accuracy, should this be:

``` suggestion
    )?.value as Caip25CaveatValue | undefined;

Reply:
https://github.com/MetaMask/metamask-extension/pull/27847/commits/7f6c67f03e16726d371d55ca6bda6826fc7fa0b0